File: 82.Remove_Duplicates_from_Sorted_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        //if(head==null) return head;
        ListNode preHead = new ListNode(0);
        preHead.next = head;
        ListNode pre = preHead, cur = head, next;
        while(cur != null) {
            boolean isDup = false;
            next = cur.next;
            while(next != null && next.val == cur.val) {
                isDup = true;
                next = next.next;
            }
            if(isDup){
                pre.next = next;
                cur = next;
            }else{
                pre = cur;
                cur = next;
            }
        }
        return preHead.next;
    }
}

File: 82_Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        preHead = ListNode(0)
        preHead.next = head
        pre, cur = preHead, head

        while cur and cur.next:
            nextN = cur.next
            if nextN.val == cur.val:
                while nextN and cur.val == nextN.val:
                    pre.next = nextN.next
                    nextN = nextN.next
                cur = nextN
            else:
                pre = cur
                cur = cur.next
        return preHead.next